flac/decoder.go

```diff
@@ -0,0 +1,109 @@
+// TODO: add seek support; only enabled in the dev branch of mewkiz/flac.
+
+package flac
+
+import (
+	"io"
+
+	"github.com/mewkiz/flac"
+	"github.com/mewkiz/flac/frame"
+	"zikichombo.org/sound"
+	"zikichombo.org/sound/freq"
+)
+
+// Decoder encapsulates state for decoding and seeking a FLAC audio stream.
+type Decoder struct {
+	stream *flac.Stream
+	frame  *frame.Frame // current frame.
+	i      int          // index of current sample in subframe(s).
+}
+
+// NewDecoder creates a decoder from a FLAC audio stream (seekable, readable).
+func NewDecoder(r io.Reader) (*Decoder, error) {
+	stream, err := flac.New(r)
+	if err != nil {
+		return nil, err
+	}
+	d := &Decoder{
+		stream: stream,
+	}
+	return d, nil
+}
+
+func (d *Decoder) Receive(dst []float64) (int, error) {
+	if len(dst)%d.Channels() != 0 {
+		return 0, sound.ErrChannelAlignment
+	}
+	n := 0                       // number of frames (samples per channel) buffered.
+	m := len(dst) / d.Channels() // number of frames (samples per channel) to buffer.
+	bps := int(d.stream.Info.BitsPerSample)
+	for n < m {
+		if d.frame == nil {
+			frame, err := d.stream.ParseNext()
+			if err != nil {
+				return n, err
+			}
+			d.frame = frame
+			d.i = 0
+		}
+		samplesLeft := len(d.frame.Subframes[0].Samples[d.i:])
+		j := m
+		if j > samplesLeft {
+			j = samplesLeft
+		}
+		if n+j > m {
+			j = m - n
+		}
+		if j == 0 {
+			panic("cannot retrieve zero samples")
```

Comment:
@wsc1 return error rather than panic here?

Reply:
I think the only possible entry into that case is if a flac file has a 0 length frame.  

I would just ignore the case and ensure any 0 length flac frame is skipped.

len(dst) != 0 because there is a check len(dst) % nC, which would panic (and maybe we should make it return an error instead of panicking)